### PR TITLE
8230623: Extract command-line help for -Xlint sub-options to new --help-lint

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch splits out the concrete keys of -Xlint into a new option, --help-lint, 
so that the output of --help-extra is kept clean and concise.

Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8230623](https://bugs.openjdk.java.net/browse/JDK-8230623): Extract command-line help for -Xlint sub-options to new --help-lint

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to fe1ece104945fa10ada9e3d4b304b36d60761968

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/1758/head:pull/1758` \
`$ git checkout pull/1758`

Update a local copy of the PR: \
`$ git checkout pull/1758` \
`$ git pull https://git.openjdk.java.net/jdk pull/1758/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1758`

View PR using the GUI difftool: \
`$ git pr show -t 1758`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/1758.diff">https://git.openjdk.java.net/jdk/pull/1758.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/man/javac.1 [hunk 1]
```diff
@@ -636,6 +636,11 @@ Specifies the format for documentation comments.
 .RS
 .RE
 .TP
+.B \f[CB]\-\-help\-lint\f[R]
+Print the supported keys of option \-Xlint.
+.RS
+.RE
+.TP
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/man/javac.1:643]: This change is somewhat problematic. The nroff man files are generated from Markdown files which are not (yet) open source. Don't ask: long story; we're working on it.

The bottom line is that this change should really be made to the upstream Markdown file.  Direct changes to this file are at risk of being overwritten at some point in the future.

**Review by jonathan-gibbons — COMMENTED** [review]
Because a new option is being added, a CSR approval is required.

**Comment by lgxbslgx** [comment]
@jonathan-gibbons Could I ask your help to create the CSR request in the bug tracker?

**Comment by lgxbslgx** [comment]
The CSR was created. Please see [JDK-8265772](https://bugs.openjdk.java.net/browse/JDK-8265772).  And I updated the copyright. Thank you for taking the time to review the CSR andthis patch.

**Comment by jbf** [comment]
@lgxbslgx I added myself as a reviewer of the CSR. `javac -X` does not output any lint options on my machine, can you update the CSR it clarifying when `-X` would or would not output lint options,  and/or remove the non-relevant occurrences of `(-X)` ?

**Comment by lgxbslgx** [comment]
@jbf  By using the current master branch code, the command `javac -X` would output the following message.

```
  --add-exports <module>/<package>=<other-module>(,<other-module>)*
        Specify a package to be considered as exported from its defining module
        to additional modules, or to all unnamed modules if <other-module> is ALL-UNNAMED.
  --add-reads <module>=<other-module>(,<other-module>)*
        Specify additional modules to be considered as required by a given module.
        <other-module> may be ALL-UNNAMED to require the unnamed module.
  --default-module-for-created-files <module-name>
        Fallback target module for files created by annotation processors, if none specified or inferred.
  -Djava.endorsed.dirs=<dirs>  Override location of endorsed standards path
  -Djava.ext.dirs=<dirs>       Override location of installed extensions
  --patch-module <module>=<file>(:<file>)*
        Override or augment a module with classes and resources
        in JAR files or directories
  -Xbootclasspath:<path>       Override location of bootstrap class files
  -Xbootclasspath/a:<path>     Append to the bootstrap class path
  -Xbootclasspath/p:<path>     Prepend to the bootstrap class path
  -Xdiags:{compact,verbose}    Select a diagnostic mode
  -Xdoclint
        Enable recommended checks for problems in javadoc comments
  -Xdoclint:(all|none|[-]<group>)[/<access>]
        Enable or disable specific checks for problems in javadoc comments,
        where <group> is one of accessibility, html, missing, reference, or syntax,
        and <access> is one of public, protected, package, or private.
  -Xdoclint/package:[-]<packages>(,[-]<package>)*
        Enable or disable checks in specific packages. Each <package> is either the
        qualified name of a package or a package name prefix followed by .*, which
        expands to all sub-packages of the given package. Each <package> can be prefixed
        with - to disable checks for the specified package or packages.
  -Xlint                       Enable recommended warnings
  -Xlint:<key>(,<key>)*
        Warnings to enable or disable, separated by comma.
        Precede a key by - to disable the specified warning.
        Supported keys are:
          all                  Enable all warnings
          auxiliaryclass       Warn about an auxiliary class that is hidden in a source file, and is used from other files.
          cast                 Warn about use of unnecessary casts.
          classfile            Warn about issues related to classfile contents.
          deprecation          Warn about use of deprecated items.
          dep-ann              Warn about items marked as deprecated in JavaDoc but not using the @Deprecated annotation.
          divzero              Warn about division by constant integer 0.
          empty                Warn about empty statement after if.
          exports              Warn about issues regarding module exports.
          fallthrough          Warn about falling through from one case of a switch statement to the next.
          finally              Warn about finally clauses that do not terminate normally.
          missing-explicit-ctor Warn about missing explicit constructors in public and protected classes in exported packages.
          module               Warn about module system related issues.
          opens                Warn about issues regarding module opens.
          options              Warn about issues relating to use of command line options.
          overloads            Warn about issues regarding method overloads.
          overrides            Warn about issues regarding method overrides.
          path                 Warn about invalid path elements on the command line.
          processing           Warn about issues regarding annotation processing.
          rawtypes             Warn about use of raw types.
          removal              Warn about use of API that has been marked for removal.
          requires-automatic   Warn about use of automatic modules in the requires clauses.
          requires-transitive-automatic Warn about automatic modules in requires transitive.
          serial               Warn about Serializable classes that do not provide a serial version ID. 
                             Also warn about access to non-public members from a serializable element.
          static               Warn about accessing a static member using an instance.
          synchronization      Warn about synchronization attempts on instances of value-based classes.
          text-blocks          Warn about inconsistent white space characters in text block indentation.
          try                  Warn about issues relating to use of try blocks (i.e. try-with-resources).
          unchecked            Warn about unchecked operations.
          varargs              Warn about potentially unsafe vararg methods.
          preview              Warn about use of preview language features.
          none                 Disable all warnings
  -Xmaxerrs <number>           Set the maximum number of errors to print
  -Xmaxwarns <number>          Set the maximum number of warnings to print
  -Xpkginfo:{always,legacy,nonempty}
        Specify handling of package-info files
  -Xplugin:"name args"
        Name and optional arguments for a plug-in to be run
  -Xprefer:{source,newer}
        Specify which file to read when both a source file and class file are found for an implicitly compiled class
  -Xprint
        Print out a textual representation of specified types
  -XprintProcessorInfo
        Print information about which annotations a processor is asked to process
  -XprintRounds
        Print information about rounds of annotation processing
  -Xstdout <filename>          Redirect standard output

These extra options are subject to change without notice.
```

Did your machine output the same message? Especially the following output:

```
  -Xlint                       Enable recommended warnings
  -Xlint:<key>(,<key>)*
        Warnings to enable or disable, separated by comma.
        Precede a key by - to disable the specified warning.
        Supported keys are:    <-------------------- Note here
          all                  Enable all warnings    <-------------------------------- Note here
          auxiliaryclass       Warn about an auxiliary class that is hidden in a source file, and is used from other files.<------Note here
          cast                 Warn about use of unnecessary casts.       <------------------------- Note here
          // <-------------- ignore many other sub-options.
  -Xmaxerrs <number>           Set the maximum number of errors to print
```

By using the code of this patch, the command `javac -X` would output the following message. We can see that the `Supported keys` is removed.

```
  --add-exports <module>/<package>=<other-module>(,<other-module>)*
        Specify a package to be considered as exported from its defining module
        to additional modules, or to all unnamed modules if <other-module> is ALL-UNNAMED.
  --add-reads <module>=<other-module>(,<other-module>)*
        Specify additional modules to be considered as required by a given module.
        <other-module> may be ALL-UNNAMED to require the unnamed module.
  --default-module-for-created-files <module-name>
        Fallback target module for files created by annotation processors, if none specified or inferred.
  -Djava.endorsed.dirs=<dirs>  Override location of endorsed standards path
  -Djava.ext.dirs=<dirs>       Override location of installed extensions
  --help-lint                  Print the supported keys of option -Xlint
  --patch-module <module>=<file>(:<file>)*
        Override or augment a module with classes and resources
        in JAR files or directories
  -Xbootclasspath:<path>       Override location of bootstrap class files
  -Xbootclasspath/a:<path>     Append to the bootstrap class path
  -Xbootclasspath/p:<path>     Prepend to the bootstrap class path
  -Xdiags:{compact,verbose}    Select a diagnostic mode
  -Xdoclint
        Enable recommended checks for problems in javadoc comments
  -Xdoclint:(all|none|[-]<group>)[/<access>]
        Enable or disable specific checks for problems in javadoc comments,
        where <group> is one of accessibility, html, missing, reference, or syntax,
        and <access> is one of public, protected, package, or private.
  -Xdoclint/package:[-]<packages>(,[-]<package>)*
        Enable or disable checks in specific packages. Each <package> is either the
        qualified name of a package or a package name prefix followed by .*, which
        expands to all sub-packages of the given package. Each <package> can be prefixed
        with - to disable checks for the specified package or packages.
  -Xlint                       Enable recommended warnings
  -Xlint:<key>(,<key>)*
        Warnings to enable or disable, separated by comma.
        Precede a key by - to disable the specified warning.
        Use option --help-lint to see the supported keys.
  -Xmaxerrs <number>           Set the maximum number of errors to print
  -Xmaxwarns <number>          Set the maximum number of warnings to print
  -Xpkginfo:{always,legacy,nonempty}
        Specify handling of package-info files
  -Xplugin:"name args"
        Name and optional arguments for a plug-in to be run
  -Xprefer:{source,newer}
        Specify which file to read when both a source file and class file are found for an implicitly compiled class
  -Xprint
        Print out a textual representation of specified types
  -XprintProcessorInfo
        Print information about which annotations a processor is asked to process
  -XprintRounds
        Print information about rounds of annotation processing
  -Xstdout <filename>          Redirect standard output

These extra options are subject to change without notice.
```

And the command `javac --help-lint` would output the following message.

```
The supported keys of option -Xlint are:
    all                  Enable all warnings
    auxiliaryclass       Warn about an auxiliary class that is hidden in a source file, and is used from other files.
    cast                 Warn about use of unnecessary casts.
    classfile            Warn about issues related to classfile contents.
    deprecation          Warn about use of deprecated items.
    dep-ann              Warn about items marked as deprecated in JavaDoc but not using the @Deprecated annotation.
    divzero              Warn about division by constant integer 0.
    empty                Warn about empty statement after if.
    exports              Warn about issues regarding module exports.
    fallthrough          Warn about falling through from one case of a switch statement to the next.
    finally              Warn about finally clauses that do not terminate normally.
    missing-explicit-ctor Warn about missing explicit constructors in public and protected classes in exported packages.
    module               Warn about module system related issues.
    opens                Warn about issues regarding module opens.
    options              Warn about issues relating to use of command line options.
    overloads            Warn about issues regarding method overloads.
    overrides            Warn about issues regarding method overrides.
    path                 Warn about invalid path elements on the command line.
    processing           Warn about issues regarding annotation processing.
    rawtypes             Warn about use of raw types.
    removal              Warn about use of API that has been marked for removal.
    requires-automatic   Warn about use of automatic modules in the requires clauses.
    requires-transitive-automatic Warn about automatic modules in requires transitive.
    serial               Warn about Serializable classes that do not provide a serial version ID. 
                         Also warn about access to non-public members from a serializable element.
    static               Warn about accessing a static member using an instance.
    synchronization      Warn about synchronization attempts on instances of value-based classes.
    text-blocks          Warn about inconsistent white space characters in text block indentation.
    try                  Warn about issues relating to use of try blocks (i.e. try-with-resources).
    unchecked            Warn about unchecked operations.
    varargs              Warn about potentially unsafe vararg methods.
    preview              Warn about use of preview language features.
    none                 Disable all warnings
```

A difference between the CSR and the output are `sub-options` and `Supported keys`. I think the name `sub-options` is good for compilers developer and the name `Supported keys` is good for users. I revised the CSR to clarify it just now .

**Comment by lgxbslgx** [comment]
FYI: The CSR was approved just now.

Could I get your help to review this patch? Thanks a lot.

**Comment by lgxbslgx** [comment]
Ping. 

This issue may block [JDK-8266239](https://bugs.openjdk.java.net/browse/JDK-8266239).
Could I ask your help to review this patch? Thanks a lot.

**Comment by lgxbslgx** [comment]
Ping.

May I ask your help to review this patch? Thanks a lot.

**Comment by lgxbslgx** [comment]
Ping again.

This issue began at Dec 2020. I would like to finish it at JDK 17.
Maybe it is good to integrate it before rampdown phase one(June 10, 2021).
What do you think about this issue? Thanks.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties [hunk 2]
```diff
@@ -172,7 +172,7 @@ javac.opt.arg.Xlint=\
 javac.opt.Xlint.custom=\
     Warnings to enable or disable, separated by comma.\n\
     Precede a key by '-' to disable the specified warning.\n\
-    Supported keys are:
+    Use option --help-lint to see the supported keys.
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties:175]: Suggest removing the word "option".
This does not affect the CSR.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties [hunk 3]
```diff
@@ -292,6 +292,10 @@ javac.opt.X=\
     Print help on extra options
 javac.opt.help=\
     Print this help message
+javac.opt.help.lint=\
+    Print the supported keys of option -Xlint
+javac.opt.help.lint.header=\
+    The supported keys of option -Xlint are:
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties:298]: Suggest: `The supported keys for -Xlint are:`
This does not affect the CSR.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties [hunk 4]
```diff
@@ -292,6 +292,10 @@ javac.opt.X=\
     Print help on extra options
 javac.opt.help=\
     Print this help message
+javac.opt.help.lint=\
+    Print the supported keys of option -Xlint
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties:296]: Suggest: `Print the supported keys for -Xlint`
This does not affect the CSR.

**Review by jonathan-gibbons — APPROVED** [review]
Approved, with some minor suggestions to improve the text in the resource file.

Thanks for hanging in there with this one.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties (hunk 2 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties]: Fixed.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties (hunk 3 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties]: Fixed.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties (hunk 4 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/resources/javac.properties]: Fixed.

**Comment by lgxbslgx** [comment]
@jonathan-gibbons Thanks for your review. I revised the resource file according to your suggestions.

/integrate

**Comment by lgxbslgx** [comment]
@jonathan-gibbons The man documentation `javac.1` needs to be revised, too. Could I get your help to revise it? Thanks a lot.

**Pull request closed without merging**